Refactor Importing Content Guides (#395)
theforeman/foreman-documentation: guides/doc-Content_Management_Guide/topics/Importing_Content.adoc

[[Importing_Content]]
== Importing Content

> This chapter outlines how you can import {customcontent} of different types to {Project}.

Reviewer: In Satellite docs, this will render as custom content of different types...

Author: Maybe 'how you can import different types of {customcontent}' sounds better?

> If you want to import Rpms, Files, or different content types to {Project}, it is done with largely the same procedures in this chapter.

Reviewer: ```suggestion
If you want to import RPMs, files, or different content types to {Project}, the underlying procedures are the same:
```

@jlsherrill when we met about Katello docs, we talked about how what was missing was context/a scenario/ a defined use case. You mentioned how users need to realise that they don't need to always do CVs, that syncing content might be enough for people. Do you think that we could create a list here of all the scenarios that importing content will serve? I can use that content as part of an introduction with a link to this section also ? Or if you prefer we can do this as part of another PR, that is fine, but since we've tackled an entire chapter in this section, it would be nice to contextualise why and who this is important to.

Author: Thats a great idea. i'll add some context!

For example, you can use the following chapters for information on specific types of {customcontent} but the underlying procedures are the same:

* xref:Managing_OSTree_Content[]
* xref:Managing_ISO_Images[]
* xref:Managing_Custom_File_Type_Content[]

[[Using_Custom_Products_in_Satellite]]
=== Using {customproducttitle}s in {Project}

ifeval::["{build}" == "satellite"]
Both Red Hat content and {customcontent} in {Project} have similarities:
endif::[]

ifeval::["{build}" != "satellite"]
Both Red Hat content and non-RH {customcontent} in {Project} have similarities:
endif::[]

* The relationship between a product and its repositories is the same and the repositories still require synchronization.
* {customproduct}s require a subscription for clients to access, similar to subscriptions to Red Hat Products.
{Project} creates a subscription for each {customproduct} you create.

For more information about creating and packaging RPMs, see the https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html-single/rpm_packaging_guide/[RPM Packaging Guide] in the Red{nbsp}Hat Enterprise Linux documentation.

> Red Hat Content is already organized into Products, while other {customcontent} can be organized into Products however you wish.  For example, Red{nbsp}Hat Enterprise Linux Server is a _Product_ in {Project}, the repositories for that product consist of different versions, different architectures, and different add-ons.

Reviewer: ```suggestion
Red Hat Content is already organized into Products, while other {customcontent} can be organized into Products however you want. For example, {RHEL} Server is a _Product_ in {Project}. The repositories for that Product consist of different versions, architectures, and add-ons.
```

For Red{nbsp}Hat repositories, products are created automatically after enabling the repository. See xref:Importing_Content-Selecting_Red_Hat_Repositories_to_Synchronize[]

[[Importing_Content-Importing_Custom_SSL_Certificates]]
=== Importing {customssltitle} Certificates

Before you create {customcontent}, ensure that you import any {customssl} certificates that you require.  This could include client certs and keys or CA certificates for the upstream repositories you will be syncing.

If you require SSL certificates and keys to download RPMs, you can add them to {Project}.

. In the {Project} web UI, navigate to *Content* > *Content Credentials*.
In the Content Credentials window, click *Create Content Credential*.
. In the *Name* field, enter a name for your SSL certificate.
. From the *Type* list, select *SSL Certificate*.
. In the *Content Credentials Content* field, paste your SSL certificate, or click *Browse* to upload your SSL certificate.
. Click *Save*.

[[Importing_Content-Importing_a_Custom_GPG_Key]]
=== Importing a {customgpgtitle} Key

> Before you create {customcontent}, ensure that you import any GPG keys that this content is signed with.  This will allow clients to validate the signature of the rpms before installation.  For Red Hat content, this is handled automatically.

Reviewer: ```suggestion
Before you create {customcontent}, ensure that you import any GPG keys that this content is signed with. Clients use these keys to validate the signature of the RPMs before installation. For Red{nbsp}Hat content, this is handled automatically.
```

.Prerequisites

> You will want a copy of the GPG key used to sign the rpm content that will be served.  Most RPM distribution providers will provide their GPG Key on their website.  You can also extract is manually from an RPM:
> 

Reviewer: ```suggestion
Ensure that you have a copy of the GPG key used to sign the RPM content that you want to use and manage in {Project}. Most RPM distribution providers provide their GPG Key on their website. You can also extract this manually from an RPM:

```

{SmartProxyServer} offers the following policies:

Immediate::
{SmartProxyServer} downloads all metadata and packages during synchronization.
Do not use this setting if the corresponding repository on {ProjectServer} is set to *On Demand* as {ProjectServer} is forced to download all the packages.

On Demand::
{SmartProxyServer} only downloads the metadata during synchronization.
{SmartProxyServer} fetches and stores packages only on the file system when directly connected clients request them.
When you use an *On Demand* download policy, content is downloaded from {ProjectServer} if it is not available on {SmartProxyServer}.

Inherit::
{SmartProxyServer} inherits the download policy for the repository from the corresponding repository on {ProjectServer}.

=== Changing the Default Download Policy

You can set the default download policy that {Project} applies to repositories that you create in all organizations.

ifeval::["{build}" == "satellite"]

Depending on whether it is a Red Hat or non-Red Hat {customrepo}, {Project} uses separate settings.
Changing the default value does not change existing settings.

endif::[]

> ifeval::["{build}" != "satellite"]

Reviewer: If the messages are the same here, we can probably drop the ifevals?

----
# hammer product list --organization $ORG --sync-plan $SYNC_PLAN
----

[[Importing_Content-Limiting_Synchronization_Speed]]

=== Limiting Synchronization Concurrency

> By default each Repository Synchronization job will fetch up to 10 files at a time.  This can be adjusted on a per repository basis.

Reviewer: ```suggestion
By default each Repository Synchronization job can fetch up to 10 files at a time.  This can be adjusted on a per repository basis.
```

Increasing the limit may improve performance, but can cause the upstream server to be overloaded or start rejecting requests.  If you are seeing Repository syncs fail due to the upstream servers rejecting requests, you may want to try lowering the limit.

> To do so from the cli:

Reviewer: ```suggestion
To do so from the CLI:
```

[options="nowrap" subs="verbatim,quotes"]
----
# hammer repository update --organization $ORG --download-concurrency 5 --id $REPO_ID
----

[[Importing_Content-Importing_a_Custom_GPG_Key]]
=== Importing a {customgpgtitle} Key

> When clients are consuming signed {customcontent}, you will want ensure the clients are configured to validate the installation of rpms with the appropriate GPG Key.  This helps to ensure that only packages from authorized sources can be installed.

Reviewer: ```suggestion
When clients are consuming signed {customcontent}, ensure that the clients are configured to validate the installation of RPMs with the appropriate GPG Key.  This helps to ensure that only packages from authorized sources can be installed.
```

> Red Hat content will already be configured with the appropriate GPG key and thus GPG Key management of Red Hat Repositories is not supported.

Reviewer: ```suggestion
Red Hat content is already configured with the appropriate GPG key and thus GPG Key management of Red Hat Repositories is not supported.
```
